finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/BufferedStream.scala

```diff
@@ -61,51 +82,46 @@ class BufferedStream(underlying: Stream, bufferCapacity: Long = 8.kilobytes.byte
           q.offer(f.asInstanceOf[Frame])
         }
         forks += q
-        val stream = if (underlying.isEmpty) {
-          Stream.empty(q)
-        } else {
-          new AsyncQueueReader { child =>
-            override protected[this] val frameQ = q
-            override def read(): Future[Frame] = bufferedStream.synchronized {
-              // If the queue is empty, pull a Frame from the underlying Stream (which will be
-              // offered to all child Streams) before polling the queue.
-              if (q.size == 0) {
-                // We must be careful to only have one pull of the underlying Stream at a time,
-                // otherwise we could build up an unbounded list of pollers.  If N forks all call
-                // read, a single pull of the underlying Stream is sufficient to satisfy all of
-                // the forks because the pulled Frame is fanned out.
-                pullState match {
-                  case Idle =>
-                    val f = bufferedStream.pull()
-                    pullState = Pulling(f)
-                    f.ensure { pullState = Idle }
-                    super.read()
-                  case Pulling(f) =>
-                    // Pull is in progress.  Try again once the pull is complete.
-                    f.before(read())
-                }
-              } else {
-                super.read()
-              }
-            }
-          }
-        }
+        val stream =
+          if (underlying.isEmpty) Stream.empty(q)
+          else new Fork(q)
         Return(stream)
       }
       case e: Exception => Throw(e)
     }
   }
 
-  private[this] sealed trait PullState
-  private[this] case object Idle extends PullState
-  private[this] case class Pulling(f: Future[Unit]) extends PullState
+  private[this] sealed trait PullState {
+    /**
+     * @return a `Future[Frame]` corresponding to the current pull attempt.
+     * @note that this is *not* thread-safe and must be synchronized on the
+     *       `BufferedStream` at the call site (as it is in the `read()`
+     *       method of forks.
+     */
+    def pullFrame(child: Fork): Future[Frame]
+  }
+
+  private[this] case object Idle extends PullState {
+    override def pullFrame(child: Fork): Future[Frame] = {
+      val f = bufferedStream.pull()
+      pullState = Pulling(f)
+      f.ensure { pullState = Idle }
+      child.readFromQueue()
```

Comment:
Now that we have to pass child in here, I don't think this really accomplishes any separation of concerns.  Since the flow of the program bounces from Fork to here and then back to Fork, I think this is actually more confusing.